model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheManager.java

```diff
@@ -109,12 +114,30 @@ public void clientRemoval(String realmId, String clientUUID, String clientId, Se
         addInvalidations(InClientPredicate.create().client(clientUUID), invalidations);
     }
 
-
     @Override
     protected void addInvalidationsFromEvent(InvalidationEvent event, Set<String> invalidations) {
         invalidations.add(event.getId());
 
         ((RealmCacheInvalidationEvent) event).addInvalidations(this, invalidations);
     }
 
+    /**
+     * Compute a cached realm and ensure that this happens only once with the current Keycloak instance.
+     * Use this to avoid concurrent preparations of a realm in parallel threads. This helps to break the load on
+     * a stampede after a server has started, were a lot of requests come in for the same realm that hasn't been cached yet.
+     * Instead of each request loading the realm in parallel, this lets the first request load the realm, and all
+     * other requests will use the cached realm, which is much more efficient.
+     */
+    public RealmAdapter computeSerialized(String id, KeycloakSession session, BiFunction<String, KeycloakSession, RealmAdapter> compute) {
+        // this locking is only to ensure that if there is a computation for the same id in the "synchronized" block below,
+        // it will have the same object instance to lock the current execution until the other is finished.
+        Object lock = cacheInteractions.computeIfAbsent(id, s -> id);
+        try {
+            synchronized (lock) {
+                return compute.apply(id, session);
```

Comment:
If we check if the cache is hot for the `id` being requested when the monitor is released, we can skip calling this code.

Doesn't it make more sense to have this logic within the `RealmCacheSession` (static CHM) instead?

Reply:
A static CHM would be shared across Keycloak session factories AFAIK. This might work in a production Keycloak, but is a pain for testing, therefore I'd rather not use it, and use a CHM per session factory instead. That's why it ended up in this class.